capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1424 @@
+    }
+
+    fn write_bytes(&self,
+                   mut write_buffer: &'static mut [u8],
+                   mut read_buffer: &'static mut [u8],
+                   recv_len: usize) {
+
+        self.set_spi_fast_mode();
+
+        self.spi.read_write_bytes(write_buffer, Some(read_buffer), recv_len);
+    }
+
+
+    fn get_response(&self, response: SDResponse, read_buffer: &[u8]) -> (u8, u8, u32) {
+
+        let mut r1: u8 = 0xFF;
+        let mut r2: u8 = 0xFF;
+        let mut r3: u32 = 0xFFFFFFFF;
+
+        // scan through read buffer for response byte
+        for i in 0..read_buffer.len() {
+            if (read_buffer[i] & 0x80) == 0x00 {
+                // status byte is always included
+                r1 = read_buffer[i];
+
+                match response {
+                    SDResponse::R2 => {
+                        // status, then read/write status. Unused in practice
+                        if i + 1 < read_buffer.len() {
+                            r2 = read_buffer[i + 1];
+                        }
+                    }
+                    SDResponse::R3 | SDResponse::R7 => {
+                        // status, then Operating Condition Register
+                        if i + 4 < read_buffer.len() {
+                            r3 = (read_buffer[i + 1] as u32) << 24 |
+                                 (read_buffer[i + 2] as u32) << 16 |
+                                 (read_buffer[i + 3] as u32) << 8 |
+                                 (read_buffer[i + 4] as u32);
+                        }
+                    }
+                    _ => {}
+                }
+
+                // response found
+                break;
+            }
+        }
+
+        (r1, r2, r3)
+    }
+
+    // updates SD card state on SPI transaction returns
+    fn process_spi_states(&self,
+                          mut write_buffer: &'static mut [u8],
+                          mut read_buffer: &'static mut [u8],
+                          _: usize) {
+
+        match self.state.get() {
+            State::SendACmd { acmd, arg } => {
+                // send the application-specific command and resume the state
+                //  machine
+                self.state.set(self.after_state.get());
+                self.after_state.set(State::Idle);
+                self.send_command(acmd, arg, write_buffer, read_buffer, 10);
+            }
+
+            State::InitReset => {
+                // check response
+                let (r1, _, _) = self.get_response(SDResponse::R1, read_buffer);
+
+                // only continue if we are in idle state
+                if r1 == 0x01 {
+                    // next send Check Voltage Range command that is only valid
+                    //  on SDv2 cards. This is used to check which SD card version
+                    //  is installed
+                    self.state.set(State::InitCheckVersion);
+                    self.send_command(SDCmd::CMD8, 0x1AA, write_buffer, read_buffer, 10);
```

Comment:
Yeah 0x1AA really is mysterious. I'll do some research on it and come up with a better comment. I believe it's just an arbitrary pattern